### DISCUSSION TITLE
Unprovable monotonicity of multiplication in bitvectors

### DISCUSSION THREAD

**Discussion by yannickmoy** [cvc5/cvc5#10084] [General]
The following VC is not provable by cvc5 1.0.5. It states that multiplication of bitvectors is monotonic when there is no overflow (which is guaranteed here by doing the multiplication of 16bits inputs in 64bits). 

```smt
(set-logic AUFBVFPDTNIRA)
(set-info :smt-lib-version 2.6)

(declare-const x (_ BitVec 16))
(declare-const y (_ BitVec 16))
(declare-const z (_ BitVec 16))

(assert
  (not
  (=>
    (bvule x y)
    (bvule (bvmul ((_ zero_extend 48) x) ((_ zero_extend 48) z)) (bvmul ((_ zero_extend 48) y) ((_ zero_extend 48) z))))))

(check-sat)
```

Is there a chance of getting cvc5 to prove such properties one day? Maybe using int-blasting or some other experimental techniques? If so, can you suggest what experiments I can make?

This is not a new problem for SMT solvers, but we'd like to investigate what's possible in the future, due to some SPARK customers having code that generates such VCs.

**Comment by yoni206** [comment]
Hi @yannickmoy , thank you for bringing this up.
Indeed int-blasting solves it quickly!

```
panda@panda:~/git/cvc5/build$ cat ~/tmp/10084.smt2
(set-logic AUFBVFPDTNIRA)
(set-info :smt-lib-version 2.6)
(declare-const x (_ BitVec 16))
(declare-const y (_ BitVec 16))
(declare-const z (_ BitVec 16))

(assert
  (not
  (=>
    (bvule x y)
    (bvule (bvmul ((_ zero_extend 48) x) ((_ zero_extend 48) z)) (bvmul ((_ zero_extend 48) y) ((_ zero_extend 48) z))))))

(check-sat)


panda@panda:~/git/cvc5/build$ time ./bin/cvc5 ~/tmp/10084.smt2 --solve-bv-as-int=iand
unsat

real    0m0.196s
user    0m0.192s
sys     0m0.004s
```

**Reply by yannickmoy** [reply]
Hi @yoni206, thanks for checking! I'm impressed. And int-blasting succeeds with more complex VCs as well such as the monotonicity of multiplication in 16bits conditioned on no overflow:

```smt
(set-logic AUFBVFPDTNIRA)
(set-info :smt-lib-version 2.6)

(declare-const x (_ BitVec 16))
(declare-const y (_ BitVec 16))
(declare-const z (_ BitVec 16))

(assert
  (not
  (=>
    (and
      (bvule x y)
      (bvule (bvmul ((_ zero_extend 48) y) ((_ zero_extend 48) z)) #x000000000000FFFF))
    (bvule (bvmul x z) (bvmul y z)))))

(check-sat)
```

or the monotonicity of multiplication conditioned on a more complex expression of no overflow:

```smt
(set-logic AUFBVFPDTNIRA)
(set-info :smt-lib-version 2.6)

(declare-const x (_ BitVec 16))
(declare-const y (_ BitVec 16))
(declare-const z (_ BitVec 16))

(assert
  (not
  (=>
    (and (bvule x y) (=> (not (= z #x0000)) (= (bvudiv (bvmul y z) z) y)))
    (bvule (bvmul x z) (bvmul y z)))))

(check-sat)
```

I see that int-blasting is still marked as experimental [in cvc5 user's doc ](https://cvc5.github.io/docs/cvc5-1.0.8/options.html). How far is it from being ready for industrial use in your opinion? Also, do you plan to integrate it in the internal strategies used by cvc5, or is it meant to remain an external option, so that we'd have to run cvc5 twice to get the best results in general?

**Reply by yannickmoy** [reply]
I'm happy to report that it also solves this challenge I presented in 2019 at [the Dagstuhl seminar on "Bringing CP, SAT and SMT together"](https://www.dagstuhl.de/seminars/seminar-calendar/seminar-details/19062):

```smt
(declare-const x (_ BitVec 32))
(declare-const y (_ BitVec 32))
(declare-const z (_ BitVec 32))
(declare-const res (_ BitVec 32))

(assert (bvult y z))
(assert (bvult x #x00010000))
(assert (bvult y #x00010000))
(assert (= res (bvudiv (bvmul x y) z)))

(assert (not (bvule res x)))
(check-sat)
```

as well as variants of the same problem with fewer constraints (e.g. remove the upper bounds on `x` and `y`) or larger bitvector sizes.

**Reply by yannickmoy** [reply]
Even better, it also solves the mix int-bv challenge I presented during the same seminar!

```smt
(declare-const x Int)
(declare-const y Int)
(declare-const u (_ BitVec 32))
(declare-const v (_ BitVec 32))

(assert (<= 0 x))
(assert (<= x 32768))
(assert (<= 0 y))
(assert (<= y 32768))

(assert (= u ((_ int2bv 32) x)))
(assert (= v ((_ int2bv 32) y)))

(assert (not (= (bvadd u v) ((_ int2bv 32) (+ x y)))))
(check-sat)
``` 

It looks like it could help with the hurdles we face when using `int2bv` currently.

**Comment by martin-cs** [comment]
IIRC cvc4's bit-vector solver used to have a "sub solver" for handling ordering problems.   It feels like quite a few of these could be handled by a light-weight relational abstraction, something like zones or DBMs?  I know a few solvers have experimented with these.  I have thought about writing one for floating-point as many operations have weak monotonicity properties.  The same code should be quite easily adapted to handle bit-vectors.

Also, as a side note, I would suggest against some of the convoluted encodings of "does not  overflow", use the overflow predicates instead.

**Reply by yannickmoy** [reply]
Hi Martin. The sub-solver looks interesting, but the VCs I sent are just the simplest examples I could imagine that show the problem of non-linear arithmetic and mixed int-bv arithmetic with the default configuration of cvc5. int-blasting seems to be a real solution here, but we need to understand how to use it. I just ran cvc5 on our complete testsuite, and I got huge regressions on:

"Error in option parsing: bv-to-int does not support quantified variables under uninterpreted functions"

So we need to understand if there is a plan to make it officially supported, and if so whether it will be integrated or not in the default configuration.

Regarding encodings of overflow, the last example is one that comes from a customer, and previous ones are simple encodings in the source language (SPARK) when someone is expressing a lemma in that language. If there was an actual need to use the overflow predicates for provability reasons, we could define a corresponding predicate in SPARK to be used in these lemmas, but that is not an identified problem at the moment.

**Comment by yoni206** [comment]
@yannickmoy 
----------------
- it is very exciting that this option can solve these problems! Thank you for sharing them.

- I am not sure at which point we decided that this is experimental, but will check whether we can make it an official option.

- As for integration in strategies, we might be able to do so. It seems that this technique is most effective on large bit-widths when bit-wise operators are not present. So there may be good reason to sometimes automatically dispatch the problem to this approach. However, on SMT-LIB problems, the approach is mostly inferior to bit-blasting.

- Can you share an example for a benchmark on which you get the error?
We decided not to support int-blasting for the combination of bit-vectors with uninterpreted functions and quantifiers, since these make the encoding more complicated, and we did not have any real example that used it and would benefit from this approach.
But if there is an application for that, we can definitely revisit this and look into supporting this combination.

@martin-cs 
-------------
did you mean one of the following options? With 10 seconds timeout (on an old version that still has these options) the original problem of this thread is not solved:

```
panda@panda:~$ timeout 10 ./cvc5-2021-05-21-x86_64-linux-opt --bv-inequality-solver ~/tmp/10084.smt2
cvc5 interrupted by SIGTERM.

panda@panda:~$ timeout 10 ./cvc5-2021-05-21-x86_64-linux-opt --bv-algebraic-solver ~/tmp/10084.smt2
cvc5 interrupted by SIGTERM.

panda@panda:~$ timeout 10 ./cvc5-2021-05-21-x86_64-linux-opt --bv-eq-solver ~/tmp/10084.smt2
cvc5 interrupted by SIGTERM.

```

**Reply by yannickmoy** [reply]
@yoni206 here is a file on which cvc5 generates this error. I did not try to minimize it, let me know if that's helpful.

[bv_as_int_error_smt2.txt](https://github.com/cvc5/cvc5/files/12820030/bv_as_int_error_smt2.txt)

**Reply by yannickmoy** [reply]
@yoni206 what do you call large bit-widths here? On the examples I sent, int-blasting is solving immediately the VC for 32bits and 64bits sizes, while cvc5 just chokes on these.

**Comment by yoni206** [comment]
Well, large is typically more than that, but when you have multiplication then problems start to occur even at 32 bits.

Thank you for the benchmark! We are looking into it.

**Comment by yoni206** [comment]
#10090 
#10091

**Reply by yoni206** [reply]
FYI both PRs are merged.

**Reply by yannickmoy** [reply]
thanks, I checked it indeed allows me to run cvc5 with int-blasting on some VCs where it crashed previously. We'll update cvc5 soon to try more globally.
